// rn-a11y-audit/constants.ts
export const touchables = [
  "TouchableOpacity",
  "TouchableHighlight",
  "TouchableWithoutFeedback",
  "Pressable",
  "Button",
  "Touchable",
];

export const ALL_ROLES = new Set<string>([
  "adjustable",
  "alert",
  "button",
  "checkbox",
  "combobox",
  "grid",
  "header",
  "image",
  "imagebutton",
  "keyboardkey",
  "link",
  "list",
  "menu",
  "menubar",
  "menuitem",
  "none",
  "progressbar",
  "radio",
  "radiogroup",
  "scrollbar",
  "search",
  "spinbutton",
  "summary",
  "switch",
  "tab",
  "tablist",
  "text",
  "timer",
  "togglebutton",
  "toolbar",
]);

// Role mapping

export const roleMap: Record<string, string> = {
  // touchables as button
  TouchableOpacity: "button",
  TouchableHighlight: "button",
  TouchableWithoutFeedback: "button",
  Pressable: "button",
  Touchable: "button",

  // native/common
  Image: "image",
  Switch: "switch",
  ActivityIndicator: "progressbar",
  FlatList: "list",
  SectionList: "list",

  // likely custom names
  Alert: "alert",
  Toast: "alert",
  Checkbox: "checkbox",
  Radio: "radio",
  RadioGroup: "radiogroup",
  Picker: "combobox",
  Dropdown: "combobox",
  Slider: "adjustable",
  Carousel: "adjustable",
  Search: "search",
  ProgressBar: "progressbar",
  Header: "header",
  Title: "header",
  Tab: "tab",
  Toolbar: "toolbar",
  Menu: "menu",
  Toggle: "togglebutton",
  Summary: "summary",
  Timer: "timer",
  Link: "link",
  Grid: "grid",
};

// State mapping

export const stateMap: Record<string, string> = {
  Switch: "checked",
  Checkbox: "checked",
  Accordion: "expanded",
  BottomSheet: "expanded",
  Radio: "selected",
  Picker: "expanded",
  Dropdown: "expanded",
  Toggle: "checked",
};

export const ariaStates = ["aria-checked", "aria-expanded", "aria-selected"];

// Error messages

export const labelCode = "@RN-AUDIT-LABEL";
export const hintCode = "@RN-AUDIT-HINT";
export const roleCode = "@RN-AUDIT-ROLE";
export const stateCode = "@RN-AUDIT-STATE";
export const nestingCode = "@RN-AUDIT-NESTING";

export const viewLabelError = `${labelCode}: Add an accessibilityLabel to your accessible View`;
export const touchableLabelError = `${labelCode}: Add an accessibilityLabel`;
export const imageLabelError = `${labelCode}: Add an accessibilityLabel and alt text to your Image`;

export const touchableHintError = `${hintCode}: Add an accessibilityHint`;

export const stateError = `${stateCode}: Add an accessibilityState`;

export const nestingError = `${nestingCode}: <View accessible> wraps interactive elements (e.g., Pressable/Touchable) and they are not reachable by screen readers. Consider removing "accessible" or moving the elements outside of the View.`
